### PR TITLE
fix(agent): emit observations for tool executions

### LINKED ISSUE #4507: [Question] graph observation是否支持trace上报tools的输入输出等？
### Question

Image
如上图所示，我利用spring-ai-alibaba-starter-graph-observation 和spring-ai-alibaba-agent-framework（1.1.2.2版本）在对接langfuse做链路检测时，必要配置项已配置完全，我的graph state日志已打印工具调用的输入输出，但是在langfuse界面未看到tools类型的span。

请问是目前graph observation暂未支持tools trace数据上报么？ 有考虑将功能加入并支持么？

### PR DESCRIPTION
**Pull request by March-77**
## Root cause

`ReactAgent` intentionally disables Spring AI's internal tool execution because `AgentToolNode` owns the execution lifecycle. As a result, tool callbacks bypassed Spring AI's `DefaultToolCallingManager` and no standard `spring.ai.tool` observation was emitted, so tracing backends could not report tool inputs and outputs.

## Changes

- wrap actual `AgentToolNode` callback execution with Spring AI's standard `ToolCallingObservationDocumentation.TOOL_CALL`
- populate `ToolCallingObservationContext` with the resolved tool definition, metadata, arguments, and result
- propagate an explicitly configured observation registry to the tool node
- inherit the registry from `CompileConfig` for starter-created agents
- retain `ObservationRegistry.NOOP` as the default
- cover both explicit and compile-config registry paths with a deterministic regression test

## User impact

Tool executions performed by graph agents now appear as standard `spring.ai.tool` observations in Micrometer/OpenTelemetry-compatible tracing backends, including tool name, input arguments, and output result. Tool execution semantics are unchanged.

## Validation

```bash
JAVA_HOME=/opt/homebrew/opt/openjdk PATH=/opt/homebrew/opt/openjdk/bin:$PATH ./mvnw -pl spring-ai-alibaba-agent-framework -Dtest=AgentToolObservationTest test
JAVA_HOME=/opt/homebrew/opt/openjdk PATH=/opt/homebrew/opt/openjdk/bin:$PATH ./mvnw -pl spring-ai-alibaba-agent-framework test
```

Result: 592 tests run, 0 failures, 0 errors, 167 skipped. Checkstyle and Spotless passed.

## Compatibility

No existing public API is changed. The tool node defaults to a no-op registry, so observation overhead is incurred only when a real registry is configured.

Fixes #4507

### CODE REVIEW

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java [hunk 1]
```diff
@@ -583,6 +597,24 @@ private ToolCallResponse executeToolCallWithInterceptors(AssistantMessage.ToolCa
...
+				.toolMetadata(toolCallback.getToolMetadata())
+				.toolCallArguments(arguments)
+				.build();
+
+		return ToolCallingObservationDocumentation.TOOL_CALL
+				.observation(null, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext, observationRegistry)
+				.observe(() -> {
+					ToolCallResponse response = execution.get();
+					observationContext.setToolCallResult(response.getResult());
+					return response;
```
[the remark this answers is not in the document]
Comment (March-77) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java]: Fixed in `845002b00`. Handled sync and async `ToolExecutionException` responses now retain the processed tool result while carrying error status, and the observation explicitly records an error whenever `response.isError()` is true. The regression verifies Micrometer receives the error callback; all 4 observation tests pass.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/DefaultBuilder.java [hunk 2]
```diff
@@ -71,10 +71,17 @@ public ReactAgent build() {
 				? getChatClientDefaultOptions(chatClient)
 				: getChatModelDefaultOptions(model);
 		ChatOptions effectiveOptions = mergeSourceOptionsWithAgentOptions(sourceOptions, this.chatOptions);
+		ObservationRegistry effectiveObservationRegistry = this.observationRegistry;
+		if (effectiveObservationRegistry == null && this.compileConfig != null) {
+			effectiveObservationRegistry = this.compileConfig.observationRegistry();
+		}
+		if (effectiveObservationRegistry == null) {
+			effectiveObservationRegistry = ObservationRegistry.NOOP;
```
[the remark this answers is not in the document]
Comment (March-77) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/DefaultBuilder.java:79]: Fixed in `845002b00`. `AgentSpecReactAgentFactory` now forwards its configured observation registry when building a ReactAgent from a supplied ChatClient. The regression builds that exact factory path and verifies the tool observation reaches the supplied registry; all 4 observation tests pass.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java [hunk 3]
```diff
@@ -583,6 +598,27 @@ private ToolCallResponse executeToolCallWithInterceptors(AssistantMessage.ToolCa
...
+				.toolCallArguments(arguments)
+				.build();
+
+		Observation observation = ToolCallingObservationDocumentation.TOOL_CALL
+			.observation(null, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext, observationRegistry);
+		return observation.observe(() -> {
+			ToolCallResponse response = execution.get();
+			observationContext.setToolCallResult(response.getResult());
+			if (response.isError()) {
+				observation.error(new IllegalStateException(response.getResult()));
```
[the remark this answers is not in the document]
Comment (March-77) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java]: Fixed in `b71dc3a4a`. Parallel observations now have a shared terminal tracker: the outer timeout marks and stops the observation when its CAS wins, while late tool completion cannot stop it again or overwrite the timeout response. A blocked synchronous tool regression verifies the timeout produces one error and both observations stop; all 6 observation tests pass.

File: spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java [hunk 4]
```diff
@@ -583,6 +598,27 @@ private ToolCallResponse executeToolCallWithInterceptors(AssistantMessage.ToolCa
...
+			Supplier<ToolCallResponse> execution) {
+		String arguments = StringUtils.hasText(request.getArguments()) ? request.getArguments() : "{}";
+		ToolCallingObservationContext observationContext = ToolCallingObservationContext.builder()
+				.toolDefinition(toolCallback.getToolDefinition())
+				.toolMetadata(toolCallback.getToolMetadata())
+				.toolCallArguments(arguments)
+				.build();
+
+		Observation observation = ToolCallingObservationDocumentation.TOOL_CALL
+			.observation(null, DEFAULT_OBSERVATION_CONVENTION, () -> observationContext, observationRegistry);
```
[the remark this answers is not in the document]
Comment (March-77) [reply spring-ai-alibaba-agent-framework/src/main/java/com/alibaba/cloud/ai/graph/agent/node/AgentToolNode.java:628]: Fixed in `b71dc3a4a`. Parallel execution captures the caller observation before dispatch and reopens its scope in each executor worker before creating the tool observation. The regression runs two parallel tools under an explicit parent and verifies both tool contexts reference that parent; all 6 observation tests pass.

**Review by oss-sentinel-ai — COMMENTED** [review]
### Summary
Reviewed the changes in this PR (660 lines changed). The changes appear to be well-structured.

### Observations
- Changes span 660 lines
- Please ensure all edge cases are covered
- Verify test coverage for the modified code paths

---
*Automated review by github-manager-bot*

**Comment by March-77** [comment]
@yuluo-yx 已重新核对当前 head、最新 main 变化、mergeability、CI、reviews 和 review threads；现有评审意见均已验证并清理，未发现需要追加代码修改的问题。烦请在方便时进行维护者审查，谢谢。
